Update health check task. (#3435)
istio/istio.io: content/docs/tasks/security/health-check/index.md

keywords: [security,health-check]
---

> Citadel has a health checking feature that can be optionally enabled.

Reviewer: ```suggestion
You can enable Citadel's health checking feature
```

> Currently, the health checking feature is able to detect the failures of Citadel CSR signing service,

Reviewer: What does CSR stand for? Please expand abbreviations when you introduce them in a document. You can use the abbreviation on the following instances.
```suggestion
to detect the failures of the Citadel CSR signing service.
```

> by periodically sending CSRs to the API and verifying the response.

Reviewer: What API are you referring to? Do you mean the reader's service? 
```suggestion
Citadel periodically sends CSRs to the API and verifies the response.
```

> Citadel contains a _prober client_ module that periodically checks the health status of Citadel's gRPC server.

Reviewer: ```suggestion
The _prober client_ module in Citadel periodically checks the health status of Citadel's gRPC server.
```

Note: because Citadel health checking currently only monitors the health status of CSR service API,
this feature is not needed if the production setup is not using the

> [SDS](/docs/tasks/security/auth-sds/) or [Mesh Expansion](/docs/setup/kubernetes/mesh-expansion/).

Reviewer: ```suggestion
[SDS](/docs/tasks/security/auth-sds/) or [Mesh Expansion](/docs/setup/kubernetes/additional-setup/mesh-expansion/).
```

## Before you begin

> * Set up Istio by following the instructions in the

Reviewer: Please introduce the list properly:
```suggestion
To complete this task, you can install Istio using one of the following paths:

* To setup Istio without using Helm, follow the instructions in the
```

>   [Kubernetes quick start](/docs/setup/kubernetes/quick-start/) with global mutual TLS enabled:

Reviewer: ```suggestion
  [Kubernetes installation guide](/docs/setup/kubernetes/install/kubernetes/). Remember to enable global mutual TLS with:
```

    {{< text bash >}}
    $ kubectl apply -f install/kubernetes/istio-demo-auth.yaml
    {{< /text >}}

>     _**OR**_

Reviewer: ```suggestion
```

>     Using [Helm](/docs/setup/kubernetes/helm-install/) with `global.mtls.enabled` set to `true`.

Reviewer: ```suggestion
 * Use [Helm](/docs/setup/kubernetes/install/helm/) to setup Istio and set the `global.mtls.enabled` flag to `true`.
```

{{< tip >}}

> You can use [authentication policy](/docs/concepts/security/#authentication-policies) to configure mutual TLS for

Reviewer: ```suggestion
Use an [authentication policy](/docs/concepts/security/#authentication-policies) to configure mutual TLS for
```

> all/selected services in a namespace (repeated for all namespaces to get global setting).

Reviewer: ```suggestion
all or only selected services in a namespace. You must repeat the policy for all namespaces to configure the setting globally.
```

See the [authentication policy task](/docs/tasks/security/authn-policy/) for details.
{{< /tip >}}

## Deploying Citadel with health checking

> Redeploy Citadel with health checking enabled.

Reviewer: ```suggestion
To enable health checking, redeploy Citadel using the configuration in `istio-citadel-with-health-check.yaml`:
```

{{< text bash >}}
$ kubectl apply -f install/kubernetes/istio-citadel-with-health-check.yaml
{{< /text >}}

## Verifying the health checker is working

Citadel will log the health checking results. Run the following in command line:

{{< text bash >}}
$ kubectl logs `kubectl get po -n istio-system | grep istio-citadel | awk '{print $1}'` -n istio-system | grep "CSR signing service"
{{< /text >}}

You will see the output similar to:

{{< text plain >}}
... CSR signing service is healthy (logged every 100 times).
{{< /text >}}

The log above indicates the periodic health checking is working.

> The health checking interval is (by default) 15 seconds. It is logged every 100 times.

Reviewer: The phrase "is logged every 100 times" is ambiguous. Do you mean: 

1. The health check is logged once every 100 checks, or
1. The health check is logged only on for the 100th, 200th, etc... check?

Assuming 1:

```suggestion
The default health checking interval is 15 seconds and is logged once every 100 checks.
```

...
{{< /text >}}

> The `liveness-probe-path` and `probe-path` are the paths to the health status file, configured at Citadel and the

Reviewer: ```suggestion
The paths to the health satus files are `liveness-probe-path` and `probe-path`. You can configure the paths in Citadel and in the
```

prober.

> The `liveness-probe-interval` is the interval to update the health status file (if Citadel is healthy).

Reviewer: ```suggestion
If Citadel is healthy, the value of the `liveness-probe-interval` entry determines the interval used to update the health status file.
```

> The `probe-check-interval` is the interval for Citadel health checking controller to call the Citadel CSR service.

Reviewer: ```suggestion
The Citadel health checking controller uses the value of the `probe-check-interval` entry to determine the interval used to call the Citadel CSR service.
```

The `interval` is the maximum time elapsed since the last update of the health status file, for the prober to consider
Citadel as healthy.

> `initialDelaySeconds` and `periodSeconds` are the initial delay and the interval between activations of the probe.

Reviewer: ```suggestion
The values in the `initialDelaySeconds` and `periodSeconds`entries determine the initial delay and the interval between each activation of the probe.
```
